crates/service/src/instrumentation.rs

```diff
@@ -0,0 +1,413 @@
+//! Service instrumentation for automatic metrics and tracing.
+//!
+//! This module provides automatic OpenTelemetry-based instrumentation for all services
+//! built with the service framework. It tracks:
+//! - Message processing metrics (count, latency, errors)
+//! - Service lifecycle metrics (launches, shutdowns)
+//! - Distributed tracing with proper span hierarchy
+//!
+//! All services automatically get:
+//! 1. Parent span wrapping entire service lifecycle
+//! 2. Child spans for launch, message processing, shutdown
+//! 3. Automatic metrics collection (counters and histograms)
+
+use std::time::Duration;
+
+use opentelemetry::{
+    global,
+    metrics::{Counter, Histogram},
+    KeyValue,
+};
+use tracing::Span as TracingSpan;
+
+/// Result of an operation (success or error).
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum OperationResult {
+    /// Operation completed successfully.
+    Success,
+    /// Operation failed with an error.
+    Error,
+}
+
+impl OperationResult {
+    /// Returns the string representation for metrics labels.
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            Self::Success => "success",
+            Self::Error => "error",
+        }
+    }
+}
+
+impl From<bool> for OperationResult {
+    fn from(success: bool) -> Self {
+        if success {
+            Self::Success
+        } else {
+            Self::Error
+        }
+    }
+}
+
+impl<T, E> From<&Result<T, E>> for OperationResult {
+    fn from(result: &Result<T, E>) -> Self {
+        match result {
+            Ok(_) => Self::Success,
+            Err(_) => Self::Error,
+        }
+    }
+}
+
+/// Reason for service shutdown.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ShutdownReason {
+    /// Normal graceful shutdown.
+    Normal,
+    /// Shutdown due to error.
+    Error,
+    /// Shutdown due to external signal.
+    Signal,
+}
+
+impl ShutdownReason {
+    /// Returns the string representation for metrics labels.
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            Self::Normal => "normal",
+            Self::Error => "error",
+            Self::Signal => "signal",
+        }
+    }
+}
```

Comment:
Same here
You should probably do `FromStr` impl and then call `as_str` in the underlying `String` to get a `&str`.